### ISSUE TITLE
[Word2Vec]: GoogleNewsModel: wrong similarity values and wordsNearest()

### ISSUE THREAD

**Issue by amouhoub** [deeplearning4j/deeplearning4j#350]
All in the title,
When I load the `GoogleNews-vectors-negative300.bin.gz` model (provided by google), the FAMOUS one I always talk about...
The similarities between two very similar words are wrong (too small values)
The TOP nearest words are wrong as well:
For the word "france":
```
[TICKER_SHOCK, Judean_desert, Remarkable_Journey, Mkulo, Bebeto, Matthew_Maniscalco, Cinema_Novo, www.TheGlen.com, Sitnik, Komatsu_Grand_Prix, Bolfing]
```
for the word "day":
```
[Inc._Nasdaq_MWRK, Deepak_Varma, Matveyeva, macular_degeneration_AMD, Antoine_Girouard, Michelangelo_Leonardo_da, NEDO, Lindsay_Dares, By_HARRY_R., Czech_Airlines_SA, Mortgage_Packagers]
```

**Comment by agibsonccc** [comment]
So looking at this a bit. I've tested the different components of this. Being:

Sorting
Cosine similarities/Dot products
Matrix offsets
Loading bigger datasets. 

I haven't seen anything significant pop out as an issue. I agree this isn't coherent.
It'd be good to have 1 or 2 sanity checks though.

**Comment by mossaab0** [comment]
I get different nearest words:

For the word "france":

    [Michael_Schwerner, Tahan, dorsi_muscle, ROT_Rally, Dr._Mechel_Paggi, MFGWatch_survey, By_Jon_Maletz, DDRM, Giant, Parvo_Virus, Khaner]

For the word "day":

    [Brand_Chocolate_Candies, Kilrush_Town, Mohammad_Nawaz_Abbasi, Barre_Montpelier_Road, +###-#-####-###, Azerbaijanis_residing, Pablo_Picasso_painting, EPRDF_TPLF, saddles, Ambrose_Barends, concrete_encasement]

**Comment by mossaab0** [comment]
More debugging:

    vec.similarity("france", "spain") == 0.6375300884246826
    vec.similarity("france", "TICKER_SHOCK") == 0.06659260392189026
    vec.similarity("france", "Michael_Schwerner") == -1.949956640601158E-9
    vec.similarity("france", "Tahan") == 1.6298145055770874E-8

Note that in the nearest list, `Michael_Schwerner` appears before `Tahan`, which is counter what their respective similarity scores indicate.

I think the problem is related to sorting.

**Comment by agibsonccc** [comment]
Ok THAT is helpful. I'll look a bit more in to how nd4j does the sorting then.

**Comment by mossaab0** [comment]
In `WordVectorsImpl`, there is this line:

    INDArray[] sorted = Nd4j.sortWithIndices(distances, 0, false);

When I follow it by this:

    System.out.println(sorted[1]);

I get huge values such as:

    [[14292.6923828125,
      523.0362548828125,
      511.9042663574219,
      502.0965881347656,
      , ..., -5592.79248046875,
      -6084.0517578125,
      -14069.369140625]]

When I comment out Arrays.sort() within Nd4j.sortWithIndices(), I get more meaningful numbers such as:

    [[0.006249852944165468,
      0.003777536563575268,
      0.007157011888921261,
      0.01655847579240799,
      , ..., 0.0039437394589185715,
      0.003572911722585559,
      1.0]]

**Comment by agibsonccc** [comment]
I'm going to replace this with using a vp/sp/kdtree instead.

https://github.com/deeplearning4j/deeplearning4j/tree/master/deeplearning4j-core/src/main/java/org/deeplearning4j/clustering

I'll still work out what's going on with sorting in nd4j though. Thanks for highlighting this. This is really weird.

**Comment by mossaab0** [comment]
Are `distances` supposed to be between 0 and 1? If this is the case, then sorting is not the problem. In fact, `range(0, distances.shape()[0]).boxed().mapToDouble(distances::getDouble).max()` returns the huge value of `14292.6923828125`. Which is consistent with sorting.

**Comment by agibsonccc** [comment]
Yes. This is why I need to look in to what's going on with nd4j. This could be anything from a referenencing error to an offset error.

**Comment by mossaab0** [comment]
These huge values are their before sorting. Sorting seems to be accurate.

**Comment by mossaab0** [comment]
I think the problem is related to normalization, and here is why:

- `WordVectorsImpl.similarity()` computes the similarity between two unit vectors.
- `WordVectorsImpl.wordsNearest()` computes the product of the word vector and the matrix vocab.

These operations give different results. In the following, I assume we have only one dimension instead of 300 (i.e., `size = 2`), that `mean` is the input word, and that `syn0` is the vocab. The outputs of the two computations are different:

    double[] d1 = new double[]{1, 1};
    double[] d2 = new double[]{1, Double.MIN_VALUE};
    INDArray mean = Nd4j.create(d1);
    INDArray vec2 = Nd4j.create(d2);
    INDArray syn0 = Nd4j.create(new double[][]{d2, d1});
    INDArray distances = syn0.mmul(mean.transpose());
    INDArray meanNorm = Transforms.unitVec(mean);
    INDArray vec2Norm = Transforms.unitVec(vec2);
    distances.diviRowVector(distances.norm2(1));
    System.out.println("WordVectorsImpl.similarity():\t" + distances.getRow(0)); // 1.0
    System.out.println("WordVectorsImpl.wordsNearest():\t" + Nd4j.getBlasWrapper().dot(meanNorm, vec2Norm)); // 0.7071067690849304

**Comment by mossaab0** [comment]
I normalized the words of each vector during `WordVectorSerializer.readBinaryModel()`. Now, the magic :-)

The top nearest words for `france` are:

    [spain, french, germany, europe, italy, england, european, belgium, usa, serbia, brazil]

Perhaps, vectors should always be stored normalized in `syn0`. Otherwise, `WordVectorsImpl.wordsNearest()` becomes expensive to compute for each call (as we need to normalize the vectors each time).

**Comment by agibsonccc** [comment]
Send a pr with some of this stuff.

**Comment by mossaab0** [comment]
Will do. But, there is a design decision that needs to be made. Should the normalization take place while loading the model, or with each call of `WordVectorsImpl.wordsNearest()`?

**Comment by agibsonccc** [comment]
Do it when the vectors are loaded.

**Comment by amouhoub** [comment]
Did a pull from master and tested on my server,
I get the following nearest for france:
````
[monopolization, rifle_marksmanship, www.conocophillips.com, Ira_Chernus, companes, Digital_Connectors, IFPE, Princes_Freeway, Runrig, officer_Tod_Leiweke, Rob__moderator]
`````
and for France (capital letter, since Google news model is case sensitive)
````
[French, Andrew_Kossover, Theater, Vladimir_Kuts, Aenlle, Harvin_migraines, burgeoning_alphabet_soup, Jokhang_temple_Ramoche_monastery, Paris, Germany, makatulong_sa]
````
Although the latter seem a little bit relevant, they are far from being the ones from Gensim's results for the same input model (google news). Besides, most of them are irrelevant at all except Paris, Germany and French
However, for the word "day", it seems fine :
`````
[week, days, morning, month, hours, afternoon, hour, weekend, evening, time, night]
`````
So my point is, the sorting or the normalization still has some issues
Mossaab, what model did you use to load w2v ? is it `GoogleNews-vectors-negative300.bin.gz` in: https://drive.google.com/file/d/0B7XkCwpI5KDYNlNUTTlSS21pQmM/edit?usp=sharing

**Comment by amouhoub** [comment]
you can try it out on my REST API
`http://www.lamsade.dauphine.fr/~mouhoub/word2vec/similarity/near?word=my_word`
just put a word instead of `my_word`in the url

**Comment by agibsonccc** [comment]
K so basically more work to do yet. I need to do the sorting stuff. Cool
thanks for verifying.

**Comment by mossaab0** [comment]
Amine, I actually had 2 PRs (375 and 376), with some time between their mergings. Did your pull include both? Especially, commenting out `distances.diviRowVector(distances.norm2(1));`?

I'm using `https://s3.amazonaws.com/dl4j-distribution/GoogleNews-vectors-negative300.bin.gz`. I believe they are identical.

**Comment by agibsonccc** [comment]
We just need to do the sorting next.
ᐧ

**Comment by amouhoub** [comment]
Mossaab, I pulled both about 5 hours after you pushed them

**Comment by amouhoub** [comment]
For some words, the issue is still, like "iPhone" : 
`````
[avoided, ZAR, WIPO, capt, altering, exonerated, adversely_affect, TEXAS, any_delays_inaccuracies, negatively_impacted, starred]
````
where it is supposed to return iPad on the top, and other related words, unlike here

**Comment by mossaab0** [comment]
Pretty strange. This is what I get. Let me check the code again.

    france	[spain, french, germany, europe, italy, england, european, belgium, usa, serbia, brazil]
    France	[French, extradites_Noriega, Belgium, Villebon_Sur_Yvette, PARIS_AFX_Gaz_de, called_Xynthia_blew, Brive_la, COLVILLE_SUR_MER, Paris, Germany, Haute_Loire_region]
    day	[week, days, morning, month, hours, afternoon, hour, weekend, evening, time, night]
    iPhone	[iPad, Iphone, IPhone, iPhone_3GS, iPhones, iPhone_3G, iOS, iphone, Palm_Pre, Apple_iPad, smartphone]

**Comment by mossaab0** [comment]
I have just done a new pull, and I get the same (correct) nearest words reported above.

**Comment by amouhoub** [comment]
Re-pulled again and still get the wrong nearest words except for "day"
I even used the file from the link you provided above, but nothing really changes

**Comment by agibsonccc** [comment]
Day: Words nearest [week, days, morning, month, hours, afternoon, hour, weekend, evening, time, night, weeks, Friday, morning_TheJournal.ie_brings, year, Brainiac_Learn_something, weekday, Monday, weekends, workday, lunchtime]


Reproduce:

git clone https://github.com/deeplearning4j/deeplearning4j-word2vec-examples
mvn clean install 
./run.sh

I'm closing this as I'm seeing good results with the latest.

**Comment by amouhoub** [comment]
The issue is not solved, it just works for the word "day" and "car" since long ago
It doesn't work for "france, France, iphone, cat, computer, etc"
```
Words nearest day: [week, days, morning, month, hours, afternoon, hour, weekend, evening, time, night, weeks, Friday, Smartest_Horse, year, Diner_Dash_Flo, Novak, Monday, RBI_single, aggravated_robbery, Airline]
Words nearest france: [monopolization, rifle_marksmanship, www.conocophillips.com, Ira_Chernus, companes, Digital_Connectors, IFPE, Princes_Freeway, Runrig, officer_Tod_Leiweke, Rob__moderator, Bishop_Fu_Tieshan, Kim_Gamel, religious_indoctrination, Soldinger, Gummi, Marc_Comeau, Lebu, kindergarteners, West_Medinipur, Cossipore]
Words nearest  France: [French, Andrew_Kossover, Theater, Vladimir_Kuts, Aenlle, Harvin_migraines, burgeoning_alphabet_soup, Jokhang_temple_Ramoche_monastery, Paris, Germany, makatulong_sa, severe_anoxic_brain, Tuv, Alexa_Conradi, stated_Barry_Polisky, SI####, Italy, continuous_slab_caster, deliciously_deadpan, Simkar_Gold_Project, Pindar_AlphaGraphics]
Words nearest iPhone: [avoided, ZAR, WIPO, capt, altering, exonerated, adversely_affect, TEXAS, any_delays_inaccuracies, negatively_impacted, starred, Neuharth, Singletary, deh, Farnsworth, NES, Streak, tailed, Flavors, Diener, downing_trees]
```

**Comment by carschno** [comment]
I have the same observations on a model I built with word2vec on German data, pasting the results:
```
deeplearning4j-word2vec-examples$ java -Xmx10g -Xms10g -cp target/deeplearning4j-word2vec-test-1.0-SNAPSHOT.jar org.deeplearning4j.App lemmatized-phrase2_size300_window10_hs1_mincount100.vec true iPad
Vector for word [-0.018865076825022697, 0.07913359254598618, -0.0206016134470701 [...]
Words nearest [Sehkraft, Geldverdienen, Rekombination, Meisterschale, Shenyang, Bayer_Leverkusen, erst_Lebenstag, genug, error, Naturschutz_Deutschland, ungeheuer_Menge, Neugier_wecken, RAF-Terrorist, studiert_Physik, Randlage, Münsterland, Vorderrad, engagiert_Bürgerin, Kunsthistoriker, Behandlungsfehler, FC_Utrecht]
```
Using the word2vec distance tool with the same model:
```
$ distance lemmatized-phrase2_size300_window10_hs1_mincount100.vec
Enter word or sentence (EXIT to break): iPad

Word: iPad  Position in vocabulary: 37158

                                              Word       Cosine distance
------------------------------------------------------------------------
                                            iPhone        0.906146
                                        iPod_touch        0.869377
                                        iPod_Touch        0.865320
                                      Apple_iPhone        0.806497
```
As you can see, the results for the nearest words are quite different. The ones yielded by distance make sense, the ones returned by App (from deeplearning4j-word2vec-examples) don't.

**Comment by agibsonccc** [comment]
This was closed because it was fixed in master

**Comment by nooralahzadeh** [comment]
Sorry same issue with master :(

**Comment by agibsonccc** [comment]
@carschno Do you have anything to back this up? This was solved with @amouhoub last I knew?

**Comment by amouhoub** [comment]
@agibsonccc @carschno It was last solved with me on an amazon server, not on my local server (university)... So the issue was related to my own system

**Comment by agibsonccc** [comment]
@nooralahzadeh That's what I thought. That's why I closed this. @nooralahzadeh I'm going to need something with your environment. I'm not going to re open this arbitrarily.

**Comment by amouhoub** [comment]
@agibsonccc you can access my machine as well if you want,,, I didn't change my access credentials,,, in case you want to compare stuff ;-)

**Comment by nooralahzadeh** [comment]
It is Fedora release 19. let me know if you need any other information

**Comment by agibsonccc** [comment]
Can you run this on AWS on another machine please? It may have to do with the binary loading or how it's being read from disk.

**Comment by nooralahzadeh** [comment]
I have just fedora machine around !

**Comment by agibsonccc** [comment]
I'm not sure what to tell you. As you can see this issue isn't something that's 100% reproducible. You're asking me to try to reproduce something inconsistently when I don't have the env to do so. As you can see from @amouhoub it has something to do with his hardware. All I can say is troubleshoot this one yourself atm. I don't know what the fix is and it's hard to reproduce across machines.  

There's plenty of alternative word2vecs you can use. I'm receptive if you'd like to try to troubleshoot this but I'm not going to chase this one down atm.

**Comment by carschno** [comment]
I am afraid I can confirm this, using either netlib-blas or jblas backend. I can't really track down the root cause at the moment either though, but it must be related to a commit during the past week, approximately.

**Comment by agibsonccc** [comment]
Ok great. I will try running this.

**Comment by agibsonccc** [comment]
So I was able to reproduce this. This was actually related to nd4j ordering. A work around for now is adding the line here before you load it.

https://github.com/deeplearning4j/deeplearning4j-word2vec-examples/blob/master/src/main/java/org/deeplearning4j/App.java#L19-19

Nothing in dl4j had changed. I had changed the default storage ordering in jblas to c ordering.

I'll work on making wordvector serializer ordering agnostic.
